File: python/2. validateSubsequence.py
nums = [3, 6, 1, -1, 0, 19, 22, 11, 7]
subNums = [6, 1, 0, 11]

# Time complexity = O(N)
# Space complexity = O(1)


def validateSubsequence(nums, sub):
    numsIndex, subIndex = 0

    while numsIndex < len(nums) and subIndex < len(sub):
        if nums[numsIndex] == sub[subIndex]:
            subIndex += 1
        numsIndex += 1

    return subIndex == len(sub)


validateSubsequence(nums, subNums)

File: js/2. validateSubsequence.js
const nums = [3, 6, 1, -1, 0, 19, 22, 11, 7]
const subNums = [6, 1, 0, 11]

// Time Complexity = O(N)
// Space Complexity = O(N)

const validateSubsequence = (nums, sub) => {
	let count = 0
	const temp = []
	for (let i = 0; i < sub.length; i++) {
		while (count <= nums.length) {
			if (nums[count] === sub[i]) {
				temp.push(nums[count])
				break
			}
			count++
		}
	}
	return temp.length === sub.length
}

// Time Complexity = O(N)
// Space Complexity = O(1)

const validateSubsequence2 = (nums, sub) => {
	let numsIndex = 0
	let subIndex = 0

	while (numsIndex < nums.length && subIndex < sub.length) {
		if (nums[numsIndex] === sub[subIndex]) subIndex++
		numsIndex++
	}
	return subIndex === sub.length
}

validateSubsequence(nums, subNums)
validateSubsequence2(nums, subNums)
